library/math/n_choose_k_mod.hpp

```diff
@@ -1,46 +1,44 @@
 /** @file */
 #pragma once
+const int N = 1001;
 /**
- * @code{.cpp}
- *     n_choose_k<1'000'000'007> nk(n); // to use `choose` with inputs strictly < n
- *     n_choose_k<1'000'003> nk(1'000'003); // to use `choose_lucas` with arbitrarily large inputs
- * @endcode
+ * nCk with arbitrarily large integers requires O(MOD) precalc
  */
-template<int MOD> struct n_choose_k {
-	vector<long long> inv, fact, inv_fact;
-	/**
-	 * @param n size
-	 * @time O(n + sqrt(MOD))
-	 * @memory O(n)
-	 */
-	n_choose_k(int n) : inv(n, 1), fact(n, 1), inv_fact(n, 1) {
-		assert(max(n, 2) <= MOD);
-		for (int i = 2; i * i <= MOD; i++) assert(MOD % i);
-		for (int i = 2; i < n; i++) {
-			inv[i] = MOD - (MOD / i) * inv[MOD % i] % MOD;
-			fact[i] = fact[i - 1] * i % MOD;
-			inv_fact[i] = inv_fact[i - 1] * inv[i] % MOD;
-		}
-	}
-	/**
-	 * @param n,k requires n < ssize(fact)
-	 * @returns number of ways to choose k objects out of n
-	 * @time O(1)
-	 */
-	long long choose(int n, int k) const {
-		if (k < 0 || n < k) return 0;
-		return fact[n] * inv_fact[k] % MOD * inv_fact[n - k] % MOD;
-	}
-	/**
-	 * @param n,k arbitrarily large integers
-	 * @returns number of ways to choose k objects out of n
-	 * @time O(log(k))
-	 */
-	long long choose_lucas(long long n, long long k) const {
-		if (k < 0 || n < k) return 0;
-		long long res = 1;
-		for (; k && k < n && res; n /= MOD, k /= MOD)
-			res = res * choose(int(n % MOD), int(k % MOD)) % MOD;
-		return res;
+long long inv[N], fact[N], inv_fact[N];
+/**
+ * @time O(n + sqrt(MOD))
+ * @memory O(n)
+ */
+template<int MOD> void calc_chooses() {
+	for (int i = 2; i * i <= MOD; i++) assert(MOD % i);
```

Comment:
there used to be an assert

`assert(max(2,N)<=MOD)`

reason for asserting N<=MOD: the code doesn't work if N>MOD, since then fact[MOD] incorrectly = 0 (as well as all higher factorials)

reason for asserting 2<=MOD: just for completeness of asserting primality

Reply:
The problem is this assert breaks our tests, so something needs to be changed. We would potentially need to switch to a variable `n` size in that case, rather than the global constant.

Reply:
first priority: make the lib code have the properties which we want (in this case the property is to RTE on user error)

second priority: keeping tests the same

Reply:
what if we change N to 1009 (a prime) and then only test choose lucas with this mod

Reply:
Yeah I agree about the priorities, I think that would be a reasonable solution